timemachine/md/hrex.py

```diff
@@ -0,0 +1,247 @@
+
+    def attempt_neighbor_swaps(
+        self,
+        neighbor_pairs: Sequence[Tuple[StateIdx, StateIdx]],
+        log_q: Callable[[ReplicaIdx, StateIdx], float],
+        n_swap_attempts: int,
+    ) -> Tuple["Hrex[_Replica]", List[Tuple[int, int]]]:
+
+        move = MixtureOfMoves([NeighborSwapMove(log_q, s_a, s_b) for s_a, s_b in neighbor_pairs])
+
+        replica_idx_by_state = list(self.replica_idx_by_state)
+        for _ in range(n_swap_attempts):
+            replica_idx_by_state = move.move(replica_idx_by_state)
+
+        fraction_accepted_by_pair = list(zip(move.n_accepted_by_move, move.n_proposed_by_move))
+
+        return Hrex(self.replicas, replica_idx_by_state), fraction_accepted_by_pair
+
+    @property
+    def state_replica_pairs(self) -> List[Tuple[StateIdx, _Replica]]:
+        return [(StateIdx(i), self.replicas[replica_idx]) for i, replica_idx in enumerate(self.replica_idx_by_state)]
+
+
+def get_cumulative_replica_state_counts(replica_idx_by_state_by_iter: Sequence[Sequence[ReplicaIdx]]) -> NDArray:
+    """Given a mapping of state index to replica index by iteration, returns an array of cumulative counts by iteration, replica, and state.
+
+    Returns
+    -------
+    NDArray
+        (iter, replica, state) -> cumulative occupancy count: int
+    """
+    replica_idx_by_state_by_iter_ = np.array(replica_idx_by_state_by_iter)  # (iter, state) -> replica
+    _, n_states = replica_idx_by_state_by_iter_.shape
+    states = np.arange(n_states)
+    replica_in_state = replica_idx_by_state_by_iter_[:, :, None] == states  # (iter, replica, state) -> bool
+    cumulative_count = np.cumsum(replica_in_state.astype(int), axis=0)  # (iter, replica, state) -> int
+    return cumulative_count
+
+
+def estimate_transition_matrix(replica_idx_by_state_by_iter: Sequence[Sequence[ReplicaIdx]]) -> NDArray:
+    """Given a mapping of state index to replica index by iteration, returns an estimate of the transition matrix.
+
+    Returns
+    -------
+    NDArray
+        (state, state) -> transition_rate: float
+    """
+    replica_idx_by_state_by_iter_ = np.array(replica_idx_by_state_by_iter)  # (iter, state) -> replica index
+    n_iters, _ = replica_idx_by_state_by_iter_.shape
+
+    # transition_by_iter: (n_iters, n_states, n_states) -> bool
+    transition_by_iter = replica_idx_by_state_by_iter_[:-1, None, :] == replica_idx_by_state_by_iter_[1:, :, None]
+
+    transition_count = np.sum(transition_by_iter, axis=0)  # (to state, from state) -> int
+    transition_rate = transition_count / n_iters  # (to state, from state) -> float
+
+    return transition_rate
+
+
+def estimate_relaxation_time(transition_matrix: NDArray) -> float:
+    """Estimate the relaxation time of permutation moves (in number of iterations) as a function of the second-largest
+    eigenvalue of the transition matrix."""
+
+    eigvals_ascending = np.linalg.eigvals(transition_matrix)
```

Comment:
Any assumptions about the underlying structure of the markov chain? Maybe assert that sum of column (or row) vectors sum to 1 before proceeding?

Reply:
Added assertion in https://github.com/proteneer/timemachine/pull/1128/commits/01755c6c64f74facea06228042af2ebfb7928470